Chessington.GameEngine/Pieces/Piece.cs

```diff
@@ -9,16 +9,28 @@ public abstract class Piece
         protected Piece(Player player)
         {
             Player = player;
+            HasMoved = false;
         }
 
         public Player Player { get; private set; }
+        public bool HasMoved { get; private set; }
 
         public abstract IEnumerable<Square> GetAvailableMoves(Board board);
 
         public void MoveTo(Board board, Square newSquare)
         {
             var currentSquare = board.FindPiece(this);
             board.MovePiece(currentSquare, newSquare);
+            HasMoved = true;
+        }
+
+        public void AddIfOnBoard(Square square, List<Square> moves)
+        {
+            if (!(square.Col < 0 | square.Col >= GameSettings.BoardSize | square.Row < 0 |
```

Comment:
You want to use `||` for or. Also having the entire expression in brackets with ! outside makes it hard to read. Can you refactor it to avoid that? 🙂